Replace markblocks with rewritten mark-blocks

andyasp opened the pull request:
#### What this PR does

Replaces `markblocks` with a rewritten version uncreatively named `mark-blocks`.

The motivation for this is that I'm in a situation (don't ask) that requires marking and then later unmarking a large number of blocks.

`markblocks` as written was limiting for this because:
-  it only supports adding marks, not removing them
-  it only supports blocks specified as an argument, which doesn't work for large lists
-  marking blocks can fail halfway (on the global marker) and then can't be retried without manual intervention
-  when a failure is encountered on a large number of blocks it would be nice to easily resume in a way that does not replicate work or require modifying the input

The rewrite addresses these points.

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

tacole02 commented on docs/internal/tools/mark-blocks.md:
```diff
@@ -0,0 +1,80 @@
+# Mark blocks tool
```
```suggestion
# Mark-blocks tool
```

andyasp replied:
Thank you for this and the other suggestions! I committed most of them and percolated them out the changes to other tools READMEs for consistency. These titles seem tricky though. "Mark-blocks" strikes me as off because it's like a mixture of the actual tool name and a title but neither. I changed it to mark-blocks here and then `mark-blocks` in the READMEs. Does that seem okay?

tacole02 commented on docs/internal/tools/mark-blocks.md:
```diff
@@ -0,0 +1,80 @@
+## `deletion` mark
+
+When `-mark-type deletion` is provided, `mark-blocks` uploads a `DeletionMark` which tells the compactor that the block provided should be deleted.
+This is a **destructive operation** (although it won't happen immediately, the default delation delay is 12h, see `-compactor.deletion-delay` value), proceed with caution.
```
```suggestion
This is a **destructive operation** (although it doesn't happen immediately, the default delation delay is 12h, see `-compactor.deletion-delay` value), proceed with caution.
```

tacole02 commented on tools/mark-blocks/README.md:
```diff
@@ -0,0 +1,63 @@
+# Mark Blocks
```
```suggestion
# Mark-blocks
```

tacole02 commented on tools/mark-blocks/README.md:
```diff
@@ -0,0 +1,63 @@
+
+## Running
+
+Running `go build .` in this directory builds the program. Then use an example below as a guide.
```
```suggestion
Run `go build .` in this directory to build the program. Then, use an example below as a guide.
```

narqo reviewed (approved):
🔥 Code-wise, looks good to me.

I'm not sure about removing the old version, though. Would it be a better for the users to leave a note that the old version is deprecated? I'd used the old `markblock` in the past, so I'd be stroked with a mild panic if I've learned the tool's gone right when I needed it next 🥲

andyasp commented:
> Code-wise, looks good to me.

Appreciate the review, thanks!

> I'm not sure about removing the old version, though.
> Would it be a better for the users to leave a note that the old version is deprecated?

These changes started as me wanting to improve this tool in-place, but by the time I was done I realized it was more of a rewrite. I renamed it to strongly signal "this has changed!" and because I think the dashed name format is more general.

As far as whether we should deprecate rather than remove, I'd much rather remove it because it's a potential footgun. When the old version fails halfway it leaves an invalid state, and rerunning results in seeing an "already marked" message rather than fixing the situation. To me, that's yet another sign that https://github.com/grafana/mimir-squad/issues/2026 should be looked into.

I added a signpost note in case a user does experience confusion. I suppose we can remove that after a release or two.
